std/experimental/checkedint.d

```diff
@@ -0,0 +1,2678 @@
+                !isUnsigned!T && hasMember!(Hook, "onOverflow"))
+        {
+            import core.checkedint;
+            static assert(is(typeof(-payload) == typeof(payload)));
+            bool overflow;
+            auto r = negs(payload, overflow);
+            if (overflow) r = hook.onOverflow!op(payload);
+            return Checked(r);
+        }
+        else
+            return Checked(mixin(op ~ "payload"));
+    }
+
+    /// ditto
+    ref Checked opUnary(string op)() return
+    if (op == "++" || op == "--")
+    {
+        static if (hasMember!(Hook, "hookOpUnary"))
+            hook.hookOpUnary!op(payload);
+        else static if (hasMember!(Hook, "onOverflow"))
+        {
+            static if (op == "++")
+            {
+                if (payload == max.payload)
+                    payload = hook.onOverflow!"++"(payload);
+                else
+                    ++payload;
+            }
+            else
+            {
+                if (payload == min.payload)
+                    payload = hook.onOverflow!"--"(payload);
+                else
+                    --payload;
+            }
+        }
+        else
+            mixin(op ~ "payload;");
+        return this;
+    }
+
+    ///
+    static if (is(T == int) && is(Hook == void)) unittest
+    {
+        static struct MyHook
+        {
+            static bool thereWereErrors;
+            static L hookOpUnary(string x, L)(L lhs)
+            {
+                if (x == "-" && lhs == -lhs) thereWereErrors = true;
+                return -lhs;
+            }
+        }
+        auto a = checked!MyHook(long.min);
+        assert(a == -a);
+        assert(MyHook.thereWereErrors);
+    }
+
+    // opBinary
+    /**
+
+    Defines binary operators `+`, `-`, `*`, `/`, `%`, `^^`, `&`, `|`, `^`, `<<`, `>>`,
+    and `>>>`. If `Hook` defines `hookOpBinary`, `opBinary` forwards to $(D
+    Checked!(typeof(hook.hookOpBinary!op(get, rhs)),
+    Hook)(hook.hookOpBinary!op(get, rhs))).
+
+    If `Hook` does not define `hookOpBinary` but defines `onOverflow`,
+    `opBinary` forwards to `hook.onOverflow!op(get, rhs)` in case an
+    overflow occurs.
+
+    If two `Checked` instances are involved in a binary operation and both
+    define `hookOpBinary`, the left-hand side hook has priority. If both define
+    `onOverflow`, a compile-time error occurs.
+
+    */
+    auto opBinary(string op, Rhs)(const Rhs rhs)
+    if (isIntegral!Rhs || isFloatingPoint!Rhs || is(Rhs == bool))
+    {
+        return opBinaryImpl!(op, Rhs, typeof(this))(rhs);
+    }
+
+    /// ditto
+    auto opBinary(string op, Rhs)(const Rhs rhs) const
+    if (isIntegral!Rhs || isFloatingPoint!Rhs || is(Rhs == bool))
+    {
+        return opBinaryImpl!(op, Rhs, typeof(this))(rhs);
+    }
+
+    private auto opBinaryImpl(string op, Rhs, this _)(const Rhs rhs)
+    {
+        alias R = typeof(payload + rhs);
+        static assert(is(typeof(mixin("payload" ~ op ~ "rhs")) == R));
```

Comment:
This `assert` is wrong for the bit shift operators. The return type for those is the type of the left operand, regardless of the signedness or size of the right operand.